northstar/src/runtime/cgroups.rs

```diff
@@ -32,244 +37,331 @@ use tokio::{
+
+        // Configure oom
+        let oom_control = path.join(OOM_CONTROL);
+        let event_control = path.join(EVENT_CONTROL);
+        let token = CancellationToken::new();
+
+        let mut event_fd = EventFd::new(0, false).expect("Failed to create eventfd");
+
+        debug!("Opening oom_control: {}", oom_control.display());
+        let oom_control = fs::OpenOptions::new()
+            .write(true)
+            .open(&oom_control)
+            .await
+            .expect("Failed to open oom_control");
+
+        debug!("Opening event_control: {}", event_control.display());
+        let mut event_control = fs::OpenOptions::new()
+            .write(true)
+            .open(&event_control)
+            .await
+            .expect("Failed to open event_control");
+        event_control
+            .write_all(format!("{} {}", event_fd.as_raw_fd(), oom_control.as_raw_fd()).as_bytes())
+            .await
+            .expect("Failed to setup event_control");
+        event_control
+            .flush()
+            .await
+            .expect("Failed to setup oom event fd");
+
+        // This task stops when the main loop receiver closes
+        let task = {
+            let stop = token.clone();
+            task::spawn(async move {
+                debug!("Listening for oom events of {}", container);
+                let mut buffer = [0u8; 16];
+
+                'outer: loop {
+                    select! {
+                        _ = stop.cancelled() => break 'outer,
+                        _ = tx.closed() => break 'outer,
+                        _ = event_fd.read(&mut buffer) => {
+                            'inner: loop {
+                                warn!("Process {} is out of memory", container);
+                                let event = Event::Container(container.clone(), ContainerEvent::CGroup(CGroupEvent::Memory(MemoryEvent {
+                                    oom: Some(1),
+                                    ..Default::default()
+                                })));
+                                match tx.try_send(event) {
+                                    Ok(_) => break 'inner,
+                                    Err(TrySendError::Closed(_)) => break 'outer,
+                                    Err(TrySendError::Full(_)) => time::sleep(time::Duration::from_millis(1)).await,
+                                }
+                            }
                         }
                     }
                 }
-            }
-            drop(event_fd);
-            drop(oom_control);
-            drop(event_control);
-            debug!("Stopped oom monitor of {}", container);
-        })
-    };
-    (stop, task)
+
+                debug!("Stopped oom monitor of {}", container);
+            })
+        };
+
+        MemoryMonitor { token, task }
+    }
+
+    /// Construct a new cgroups v2 memory monitor
+    async fn new_v2(container: Container, path: &Path, tx: EventTx) -> MemoryMonitor {
```

Comment:
It feels like some duplication in these functions but they are different enough to not make it obvious >:(

Reply:
Yes. Merging those two would remove trivial code add add fuzz ;-)